src/dbnode/storage/flush.go

```diff
@@ -102,28 +105,69 @@ func (m *flushManager) Flush(
 
 	defer m.setState(flushManagerIdle)
 
-	// create flush-er
-	flushPersist, err := m.pm.StartFlushPersist()
+	namespaces, err := m.database.GetOwnedNamespaces()
 	if err != nil {
 		return err
 	}
 
-	namespaces, err := m.database.GetOwnedNamespaces()
+	// Perform three separate loops through all the namespaces so that we can
+	// emit better gauges, i.e. all the flushing for all the namespaces happens
+	// at once, then all the cold flushes, then all the snapshotting. This is
+	// also slightly better semantically because flushing should take priority
+	// over cold flushes and snapshotting.
+	//
+	// In addition, we need to make sure that for any given shard/blockStart
+	// combination, we attempt a flush and then a cold flush before a snapshot
+	// as the snapshotting process will attempt to snapshot any unflushed blocks
+	// which would be wasteful if the block is already flushable.
+	multiErr := xerrors.NewMultiError()
+	if err = m.dataWarmFlush(namespaces, tickStart, dbBootstrapStateAtTickStart); err != nil {
+		multiErr = multiErr.Add(err)
```

Comment:
This actually needs to either be a hard error, or if it happened, block cold flushes because if warm flushes fail the cold flushes will attempt to flush blocks that have not yet been warm flushed which violates the model that cold flushes for a block never happen before a warm flush

Reply:
...also I just realized that the `WarmFlush` method has this logic:

```golang
shardBootstrapStateBeforeTick, ok := shardBootstrapStatesAtTickStart[shard.ID()]
		if !ok || shardBootstrapStateBeforeTick != Bootstrapped {
			// We don't own this shard anymore (!ok) or the shard was not bootstrapped
			// before the previous tick which means that we have no guarantee that all
			// bootstrapped blocks have been rotated out of the series buffer buckets,
			// so we wait until the next opportunity.
			n.log.
				With(zap.Uint32("shard", shard.ID())).
				With(zap.Any("bootstrapStateBeforeTick", shardBootstrapStateBeforeTick)).
				With(zap.Bool("bootstrapStateExists", ok)).
				Debug("skipping snapshot due to shard bootstrap state before tick")
			continue
		}
```

but the `ColdFlush` logic does not which means that cold flush could end up trying to flush a block that was never WarmFlushed.

Can we just add a blanket check inside of the `ColdFlush` method itself to make sure that a warm flush has succeeded before attempting a cold flush for any blocks?

Reply:
Yeah that makes sense. We should just check the shard flush state since that is the authoritative location on whether something got successfully flushed or not.

Reply:
ok so this may still actually need to be a hard error. I was thinking through it and even with your check in the shard you could still end up in a bad spot:

1. Warm flush fails
2. Cold flush skips block cause it wasn't warm flushed (but looks like it succeeded cause you just skipped it)
3. Snapshot completes successfully
4. Commitlogs containing unflushed cold writes get deleted <- durability lost here

I think you need to make this a hard error which solves one problem. The next problem is that the WarmFlush thing can choose to skip flushes (without returning an error) because of the `shardBootstrapStateBeforeTick` thing. We can fix that by making it return an error (and it'll happen on the next go around) or by removing it entirely because I Don't think its required now that we load blocks directly into the buffer on bootstrap and flush out of the buffer too.

Lets not make that last change in this P.R though, just file an issue for it.

My guess is that we also want to make the ColdFlush logic not skip blocks that haven't been WarmFlushed yet, but actually return an error if that situations happens because it means we messed up this lifecycle and should go fix the bug

Reply:
(invariant violated error)